File: MLR-spring/src/main/java/com/mlr/controller/WalletController.java
package com.mlr.controller;

import com.mlr.dto.AddMoneyDto;
import com.mlr.dto.PaymentDto;
import com.mlr.dto.TransactionDto;
import com.mlr.dto.WalletBalanceDto;
import com.mlr.service.WalletService;
import jakarta.validation.Valid;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/wallet")
@PreAuthorize("hasRole('USER')")
public class WalletController {

    @Autowired
    private WalletService walletService;

    @GetMapping("/balance")
    public ResponseEntity<WalletBalanceDto> getBalance(Authentication authentication) {
        String userId = authentication.getName();
        return ResponseEntity.ok(walletService.getBalance(userId));
    }

    @PostMapping("/add-money")
    public ResponseEntity<?> addMoney(@Valid @RequestBody AddMoneyDto addMoneyDto, Authentication authentication) {
        if (addMoneyDto.getAmount() <= 0) {
            return ResponseEntity.badRequest().body("Amount must be positive");
        }

        String userId = authentication.getName();
        WalletBalanceDto result = walletService.addMoney(userId, addMoneyDto);
        return ResponseEntity.ok(result);
    }

    @PostMapping("/pay")
    public ResponseEntity<?> pay(@Valid @RequestBody PaymentDto paymentDto, Authentication authentication) {
        if (paymentDto.getAmount() <= 0) {
            return ResponseEntity.badRequest().body("Amount must be positive");
        }

        try {
            String userId = authentication.getName();
            WalletBalanceDto result = walletService.pay(userId, paymentDto);
            return ResponseEntity.ok(result);
        } catch (IllegalArgumentException | IllegalStateException ex) {
            return ResponseEntity.badRequest().body(ex.getMessage());
        }
    }

    @GetMapping("/transactions")
    public ResponseEntity<List<TransactionDto>> getTransactions(
            @RequestParam(defaultValue = "10") int limit, 
            Authentication authentication) {
        
        String userId = authentication.getName();
        return ResponseEntity.ok(walletService.getRecentTransactions(userId, limit));
    }
}

File: MLR/WalletController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using MLR.DTOs;
using MLR.Services;

namespace MLR.Controllers;

[ApiController]
[Route("api/[controller]")]
[Authorize(Roles = "USER")]
public class WalletController : ControllerBase
{
    private readonly IWalletService _walletService;

    public WalletController(IWalletService walletService)
    {
        _walletService = walletService;
    }

    [HttpGet("balance")]
    public async Task<IActionResult> GetBalance()
    {
        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var balance = await _walletService.GetBalanceAsync(userId);
        return Ok(balance);
    }

    [HttpPost("add-money")]
    public async Task<IActionResult> AddMoney([FromBody] AddMoneyDto addMoneyDto)
    {
        if (addMoneyDto.Amount <= 0)
            return BadRequest("Amount must be positive");

        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var result = await _walletService.AddMoneyAsync(userId, addMoneyDto);
        return Ok(result);
    }

    [HttpPost("pay")]
    public async Task<IActionResult> Pay([FromBody] PaymentDto paymentDto)
    {
        if (paymentDto.Amount <= 0)
            return BadRequest("Amount must be positive");

        try
        {
            var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
            var result = await _walletService.PayAsync(userId, paymentDto);
            return Ok(result);
        }
        catch (InvalidOperationException ex)
        {
            return BadRequest(ex.Message);
        }
        catch (ArgumentException ex)
        {
            return BadRequest(ex.Message);
        }
    }

    [HttpGet("transactions")]
    public async Task<IActionResult> GetTransactions([FromQuery] int limit = 10)
    {
        var userId = User.FindFirst(ClaimTypes.NameIdentifier)?.Value!;
        var transactions = await _walletService.GetRecentTransactionsAsync(userId, limit);
        return Ok(transactions);
    }
}